presto-hive/src/main/java/com/facebook/presto/hive/HiveZeroRowFileCreator.java

```diff
@@ -75,15 +85,55 @@ public void createFiles(ConnectorSession session, HdfsContext hdfsContext, Path
         }
     }
 
-    private static void writeZeroRowFile(ConnectorSession session, Path target, JobConf conf, Properties properties, String serDe, String outputFormatName)
+    private byte[] generateZeroRowFile(
+            ConnectorSession session,
+            HdfsContext hdfsContext,
+            Properties properties,
+            String serDe,
+            String outputFormatName,
+            HiveCompressionCodec compressionCodec)
     {
-        // Some serializers such as Avro set a property in the schema.
-        initializeSerializer(conf, properties, serDe);
+        String tmpDirectoryPath = System.getProperty("java.io.tmpdir");
+        String tmpFileName = format("presto-hive-zero-row-file-creator-%s-%s", session.getQueryId(), randomUUID().toString());
+        java.nio.file.Path tmpFilePath = Paths.get(tmpDirectoryPath, tmpFileName);
 
-        // The code below is not a try with resources because RecordWriter is not Closeable.
-        RecordWriter recordWriter = HiveWriteUtils.createRecordWriter(target, conf, properties, outputFormatName, session);
         try {
+            Path target = new Path(format("file://%s/%s", tmpDirectoryPath, tmpFileName));
+            JobConf conf = configureCompression(hdfsEnvironment.getConfiguration(hdfsContext, target), compressionCodec);
+
+            // Some serializers such as Avro set a property in the schema.
+            initializeSerializer(conf, properties, serDe);
+
+            // The code below is not a try with resources because RecordWriter is not Closeable.
+            RecordWriter recordWriter = HiveWriteUtils.createRecordWriter(
+                    target,
+                    conf,
+                    properties,
+                    outputFormatName,
+                    session);
             recordWriter.close(false);
+
+            return readAllBytes(tmpFilePath);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        finally {
+            if (exists(tmpFilePath)) {
```

Comment:
This exists check might not be needed because we could just try to delete the file regardless if it exists, but up to you.

Reply:
It the file doesn't exist, the `Files.delete` method will throw the `NoSuchFileException` (https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#delete(java.nio.file.Path)).

I just realized that there's the `deleteIfExists` method though. Let me use it.